### ISSUE TITLE
Controlling the map via queryparams (and animation issues)

### ISSUE THREAD

**Issue by timmyomahony** [miguelcobain/ember-leaflet#48]
Hey, 

I have a fullscreen map on a route. The route has `lat`, `lng` and `zoom` query params used to control the map (when these change, a model refresh is fired and markers are grabbed from the server). I update these controller properties when the user moves the map.

Roughly, I have: 

``` javascript
export default Ember.Controller.extend({
  lat: null,
  lng: null,
  zoom: null,
  actions: {
    updatePosition(e) {
        var map = e.target;
        this.setProperties({
          lat: map.getCenter().lat,
          lng: map.getCenter().lng,
          zoom: map.getZoom()
        });
    },
    markerClicked(marker) {
       this.transitionToRoute('my.route', {
        queryParams: {
          lat: marker.get('lat'),
          lng: marker.get('lng'),
          zoom: 16
        }
      });
    },
  }
});
```

and

```
{{#leaflet-map lat=lat lng=lng zoom=zoom onMoveend=(action "updatePosition")}}
  {{tile-layer url="https://cartodb-basemaps-{s}.global.ssl.fastly.net/light_all/{z}/{x}/{y}.png"}}
    {{#each model as |place|}}
      {{marker-layer ... onClick=(action "markerClicked" place)}}
    {{/each}}
{{/leaflet-map}}
```

This is actually working really nicely, but there is a slight bug. When I use `transitionToRoute` to change the position of the map via query params, it seems that the component is firing two events and one is cancelling the other out. The result is that when I transition, the zoom animation will work but the pan animation won't fire properly. Instead, the map will move a little bit and I will have to touch the map again to get the marker to center correctly.

I have temporarily fixed this by simply disabling the zoom animation: 

```
{{#leaflet-map lat=lat lng=lng zoom=zoom zoomAnimation=false onMoveend=(action "updatePosition")}}
```

and now it works nicely but I was wondering is there anything that can be done to get both the animations working together? 

**Disclaimer**
I have simplified the problem a bit. I know that I could simply change the `lat` `lng` properties when I click the marker and that would update the map component automatically but in my actual application the route with the map is actually a parent to another route which I am transitioning to so I have to use query params

**Comment by miguelcobain** [comment]
Can you debug what triggers the first and second moveend event?
I don't see why two movend should trigger. Sometimes when you click on a marker, a popup opens and the map moves to include that popup. This would trigger an extra moveend indeed.

**Comment by timmyomahony** [comment]
Hey, sorry for the replay on this. I think might actually just have to do with setting both the lat/lng and the zoom at the same time. 

I notice if I do this in the action handler:

```
this.setProperties({
  lat: map.getCenter().lat,
  lng: map.getCenter().lng,
  zoom: map.getZoom()
});
```
the map gets stuck. When I drag it again it will correct itself. If I just do:

```
this.setProperties({
  lat: map.getCenter().lat,
  lng: map.getCenter().lng
});
```
the map pans OK. So maybe it's something to do with two separate animations (a zoom and pan) cancelling each other out? I'll have to do more investigating

**Comment by miguelcobain** [comment]
A failing test would be awesome.

**Comment by pedrokost** [comment]
This affects me as well. I have a map on a parent route. When the child route loads, it changes the map center and zoom. The zooming works, but the map is not correctly centered. Removing the zoom change makes the centering work properly.

**Comment by pedrokost** [comment]
I was able to resolve my issue, although I am not fully certain what fixed it (I could not reproduce the issue in a brand new app)

The bug was present when:
1- A model was loaded asynchronously in the child router
2- In the after model hook I had a (blank) afterRender call
3- The leaflet map had maxBounds defined

Eliminating the 2 and 3 seemed to resolve this issue in my app. But again, I wasn't able to reproduce the issue in a test app.

**Comment by jozefvaclavik** [comment]
Having same issue. If I do
```
      this.setProperties({
        latitude: response.meta.coordinates[0],
        longitude: response.meta.coordinates[1],
        zoom: 6
      });
```
the map only zooms in to previously set `latitude`/`longitude`. If I remove `zoom: 6`, map centers correctly.

**Comment by delkano** [comment]
I have this exact same issue, and after hours trying to bypass it by all workarounds I could think of and obtaining zero results I decided to submit a bug.

I've composed a minimalistic Twiddle which I believe illustrates the issue. Since you can play with all three values, you can see that only the zoom changes when you try to update all three. However, if you comment the `this.set("zoom" this.newZoom)` line, then the map pans as it should.

https://ember-twiddle.com/60b53c1d647021aaab3639786d11d804?openFiles=controllers.application.js%2C

I think this happens because, whenever you update all three values, zoom is called first, which triggers a "moveend" event, which updates the lat and lng values, which cancels the panning.

**Comment by delkano** [comment]
After testing a bit around. I've found a way to solve this. It's not beautiful, so I prefer not to submit it as a PR, but it can solve things for people with this issue.

The problem is that leaflet-ember does define two different observers, one for `zoom` (calling `setZoom`) and one for `center` (calling `panTo`). When both properties are changed in the same cycle, one is called before, then the map starts its animation, calls `onmoveend`, which resets the second property, and so the second observer is never called (or does nothing).

What I've done is gone into the `leaflet-map.js` addon code, and changed the `leafletProperties` array to this:

    leafletProperties: Object.freeze([
      /*'zoom:setZoom:zoomPanOptions',*/ 'minZoom', 'maxZoom',
      /*'center:panTo:zoomPanOptions',*/
      'bounds:fitBounds:fitBoundsOptions', 'maxBounds'
    ]),
    flyToNewPlace: observer('zoom', 'center', function() {
      this._layer.flyTo(this.center, this.zoom, this.zoomPanOptions);
    }),

In other words, I've removed the individual `zoom` and `center` observers and replaced them with one unified observer which calls `flyTo` (a Leaflet map method which animates both center and zoom simultaneously).

**Comment by miguelcobain** [comment]
Please reopen if issue persists on master.
